File: LeetCode/2542. 最大子序列的分数.java
class Solution {
     public long maxScore(int[] nums1, int[] nums2, int k) {
        long maxScore = 0;
        int n = nums1.length;
        int[][] matrix = new int[n][2];
        for (int i = 0; i < n; i++) {
            matrix[i][0] = nums1[i];
            matrix[i][1] = nums2[i];
        }
        Arrays.sort(matrix,(a,b) -> b[1] - a[1]);
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        long sum = 0;
        for (int i = 0; i < n; i++) {
            if (i >= k) {
                sum -= pq.poll();
            }
            int[] cur = matrix[i];
            sum += cur[0];
            pq.offer(cur[0]);
            if (i >= k - 1) {
                maxScore = Math.max(maxScore, sum * cur[1]);
            }
        }
        return  maxScore;
    }
}

File: LeetCode/2542. 最大子序列的分数.cpp
class Solution {
public:
    long long maxScore(vector<int>& nums1, vector<int>& nums2, int k) {
                    int size = nums1.size();
            vector<pair<int,int>> olds;
            for (int i = 0; i < size; i++)
            {
                olds.push_back(pair<int,int>(-nums2[i],i));
            }
            sort(olds.begin(),olds.end());
            priority_queue<int,vector<int>,greater<int>> pq;
            long long sum = 0;
            for (int i = 0; i < k; i++)
            {
                int j = olds[i].second;
                sum += nums1[j];
                pq.push(nums1[j]);
            }
            long long ans = 0;
            ans = nums2[olds[k - 1].second] * sum;
            for (int i = k; i < size; i++)
            {
                int min = pq.top();pq.pop();
                sum -= min;
                int j = olds[i].second;
                sum += nums1[j];
                pq.push(nums1[j]);
                ans = max(ans, sum* nums2[j]);
            }
             
            return ans;
    }
};